File: src/cpf/cpf-validator.ts
export class CPFValidator {
  public static isValid(cpf: string): boolean {
    const digits = cpf.replace(/\D/g, '');
    if (digits.length !== 11) return false;
    if (/^(\d)\1{10}$/.test(digits)) return false;

    let sum = 0;
    for (let i = 0; i < 9; i++) sum += Number(digits[i]) * (10 - i);
    let check = (sum * 10) % 11;
    if (check === 10) check = 0;
    if (check !== Number(digits[9])) return false;

    sum = 0;
    for (let i = 0; i < 10; i++) sum += Number(digits[i]) * (11 - i);
    check = (sum * 10) % 11;
    if (check === 10) check = 0;
    return check === Number(digits[10]);
  }
}

File: src/main/java/org/project/mechanic_shop/cpflogin/CpfValidator.java
package org.project.mechanic_shop.cpflogin;

public final class CpfValidator {

	private CpfValidator() {
	}

	public static boolean isValid(String rawDocument) {
		if (rawDocument == null) {
			return false;
		}

		String cpf = rawDocument.replaceAll("\\D", "");
		if (cpf.length() != 11 || cpf.chars().distinct().count() == 1) {
			return false;
		}

		int firstCheckDigit = calculateCheckDigit(cpf.substring(0, 9), 10);
		int secondCheckDigit = calculateCheckDigit(cpf.substring(0, 9) + firstCheckDigit, 11);

		return cpf.equals(cpf.substring(0, 9) + firstCheckDigit + secondCheckDigit);
	}

	private static int calculateCheckDigit(String base, int startWeight) {
		int sum = 0;
		int weight = startWeight;
		for (char c : base.toCharArray()) {
			sum += (c - '0') * weight;
			weight--;
		}

		int remainder = sum % 11;
		return remainder < 2 ? 0 : 11 - remainder;
	}
}
